Add aa v0.7.0

oneleo opened the pull request:
# Description
- Rename old aa package to aa6
- Note: Unexpected address changes in SimpleAccount and PasskeyAccount during deploying.

# Setup container

```shell
$ forge install && make testnet-up
# ...
$ make testnet-down
```

# Foundry test

```shell
$ forge fmt && forge build && forge test
```

oneleo committed:
Remove submodule lib/account-abstraction

oneleo committed:
Set submodules to specific versions v0.6.0 and v0.7.0

oneleo committed:
chg: Rename old aa package to aa6

Note: Unexpected address changes in SimpleAccount and
PasskeyAccount during deploying.

oneleo committed:
Remove submodule

oneleo committed:
refac: Add submodules

1. Add submodule account-abstraction v0.6.0, v0.7.0
2. Add submodule openzeppelin-contracts v4.9.6, v5.0.2

oneleo committed:
chg: Add v0.7.0 contract

1. forge build passes
2. forge test for v0.7.0 not yet passing

oneleo committed:
chg: Remove PasskeyAccount contract related to v0.7.0

1. Remove v0.7.0-related PasskeyAccount contract to focus on v0.7.0
   planning.

oneleo committed:
chg: Update address in README.md and path in entry.sh

oneleo committed:
chg: Update path in deployment.md

pilagod commented:
I would be favor of organizing directories first by function/intention, and then implementation details.

So I would suggest to cut them like the following:

* lib/account-abstraction/{v0.6, v0.7}
* lib/openzeppelin-contracts/{v4, v5}
* src/account/{v0.6, v0.7}
* test/account/{v0.6, v0.7}

And since interfaces are indifferent to versions, it doesn't need to cut further.

And I think remapping as `@aa6` is not an ideal attempt. I would suggest to remap at most to `@account-abstraction=lib/account-abstraction` and uses `@account-abstraction/v0.6/...` to import

oneleo committed:
chg: Simplify the foundry remappings path

oneleo committed:
chg: Remove submodules

oneleo committed:
chg: Change the submodule pathes

oneleo committed:
adjust and format the test

pilagod commented on src/artifacts/0.6/Artifacts.sol:
```diff
@@ -0,0 +1,10 @@
+import { VerifyingPaymaster } from
+    "@aa/0.6/contracts/samples/VerifyingPaymaster.sol";
+
+contract EntryPoint0p6 is EntryPoint { }
```
When deploying we can specify which version of entry point to deploy, I think we don't need to do this.

oneleo replied:
This adjustment is used only in [setUp()](https://github.com/pilagod/waallet-contract/pull/29/files#diff-bc65220baf3b6c5333fafaac8b448bfbbd0b8f3e44ba232a268885ceaaf856afR36) in the foundry test because `vm.getCode()` cannot recognize different contract versions with the same name.

Or it triggers the `[FAIL. Reason: setup failed: Multiple matching artifacts found]` error.

<img width="802" alt="截圖 2024-06-05 上午11 21 17" src="https://github.com/pilagod/waallet-contract/assets/7114046/85bc77d6-b4d3-4a3e-b09f-02513703427b">

pilagod replied:
Couldn't it get code by specific path?

pilagod replied:
Another work around is to deploy EntryPoint first to a random address, and get code from it to etch to our favorable address.

pilagod commented on src/account/0.6/PasskeyAccount.sol:
```diff
@@ -1,10 +1,13 @@
+    UserOperation
+} from "@aa/0.6/contracts/interfaces/IEntryPoint.sol";
+import { IPasskeyAccount, Passkey } from "../../interfaces/IPasskeyAccount.sol";
+import { SimpleAccount } from "@aa/0.6/contracts/samples/SimpleAccount.sol";
```
I would like to make library import more comprehensive

`@aa` -> `@account-abstraction`
`@oz` -> `@openzeppelin-contracts`

oneleo replied:
OK, the `@openzeppelin` remappings need to be moved down in `foundry.toml` because they will conflict with `lib/account-abstraction/0.6 (or 0.7):@openzeppelin`.

pilagod replied:
Why `@openzeppelin-contracts` would conflict with `lib/account-abstraction/0.6:@openzeppelin`?

oneleo replied:
Sorry, I misunderstood. `@openzeppelin-contracts` doesn't conflict with `@openzeppelin`.

pilagod commented on test/0.6/PasskeyAccount.t.sol:
```diff
@@ -5,13 +5,15 @@ pragma solidity ^0.8.13;
 import { Test } from "forge-std/Test.sol";
```
Test structure should follow source code. It should be `test/account/0.6/PasskeyAccount`

oneleo committed:
chg: Adjust the remapping name

1. Change remappings from short to full names.
2. Move the foundry test to the correct directory.
3. Move `@openzeppelin` remappings down in `foundry.toml` to avoid
   conflict with `lib/account-abstraction/0.6 (or 0.7):@openzeppelin`.

oneleo committed:
chg: Use entryPointEntity to get the bytecode instead of vm.getCode()

oneleo committed:
chg: Change remappings from short to full names

oneleo committed:
chg: Get entryPoint bytecode via type(Contract).creationCode

oneleo committed:
chg: Update address in README.md and path in entry.sh

oneleo committed:
chg: Split contract operator and deployer roles

oneleo committed:
chg: Update the deployed address in README.md

pilagod commented on README.md:
```diff
@@ -32,21 +32,28 @@ make testnet-up
-| PasskeyAccountFactory | 0x5FC8d32690cc91D4c39d9d3abcBD16989F875707 |                               |
-| PasskeyAccount        | **_Depending on PASSKEY env variables_**   | Balance: 100 ether            |
-| VerifyingPaymaster    | 0x2279B7A0a67DB372996a5FaB50D91eAA73d2eBe6 | EntryPoint deposit: 100 ether |
+- Deployer 1: Non-account abstraction contracts
```
```suggestion
#### Periphery contracts
```

pilagod commented on README.md:
```diff
@@ -32,21 +32,28 @@ make testnet-up
+| Counter  | 0x8464135c8F25Da09e49BC8782676a84730C318bC |
 
-The owner of these contracts is `0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266`, which is the first address derived from testing mnemonic `test test test test test test test test test test test junk`, and it has unlimited balance of ether.
+- Deployer 2: Account abstraction v0.6.0 contracts
```
```suggestion
#### Account abstraction v0.6.0 contracts
```

pilagod commented on README.md:
```diff
@@ -32,21 +32,28 @@ make testnet-up
+
+> If you use default passkey, the PasskeyAccount address would be `0xF4bb6e38fC8A5ec977D4Fdc74B4E0fa84c8dc704`.
+
+These contracts are mainly owned by `0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266`. For easier management, deployer 1 (`0x70997970C51812dc3A010C7d01b50e0d17dc79C8`) handles Counter and non-account abstraction contracts. Deployer 2 (`0x3C44CdDdB6a900fa2b585dd299e03d12FA4293BC`) handles account abstraction v0.6.0 contracts. Deployer 3 (`0x90F79bf6EB2c4f870365E785982E1f101E93b906`) handles account abs ...
```
```suggestion
> [!NOTE]
> These contracts are mainly owned by `0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266`. 

For management purpose, we use deployer 1 (`0x70997970C51812dc3A010C7d01b50e0d17dc79C8`) to deploy [periphery contracts](#periphery-contracts). Deployer 2 (`0x3C44CdDdB6a900fa2b585dd299e03d12FA4293BC`) deploys [account abstraction v0.6.0 contracts](#account-abstraction-v0.6.0-contracts). Deployer 3 (`0x90F79bf6EB2c4f870365E785982E1f101E93b906`) deploys [account abstraction v0.7.0 contracts](#account-abstraction-v0.7.0-contracts). Contract owner and deployers are derived from the first four addresses of the mnemonic `test test test test test test test test test test test junk`, and the owner (`0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266`) has unlimited balance of ether.
```

pilagod commented on README.md:
```diff
@@ -32,21 +32,28 @@ make testnet-up
+
+These contracts are mainly owned by `0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266`. For easier management, deployer 1 (`0x70997970C51812dc3A010C7d01b50e0d17dc79C8`) handles Counter and non-account abstraction contracts. Deployer 2 (`0x3C44CdDdB6a900fa2b585dd299e03d12FA4293BC`) handles account abstraction v0.6.0 contracts. Deployer 3 (`0x90F79bf6EB2c4f870365E785982E1f101E93b906`) handles account abs ...
+
+> Please make sure the contracts in the submodules(`lib`) are added into `src/artifacts/0.6/Artifacts.sol` or `src/artifacts/0.6/Artifacts.sol` for being compiled.
```
```suggestion
> Please make sure the contracts in the submodules(`lib`) are added into `src/artifacts/{0.6, 0.7}/Artifacts.sol` for being compiled.
```

oneleo replied:
Considering shell compatibility, would it be better to use `src/artifacts/{0.6,0.7}/Artifacts.sol` (without spaces in between)?

pilagod commented on docker/testnet/geth_setup/entry.sh:
```diff
@@ -7,29 +7,55 @@ rpc_url="http://geth:8545"
 operator_address="0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266"
 operator_private_key="0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80"
 
+# For deploying Counter and other contracts
+deployer_1_address="0x70997970C51812dc3A010C7d01b50e0d17dc79C8"
+deployer_1_private_key="0x59c6995e998f97a5a0044966f0945389dc9e86dae88c7a8412f4603b6b78690d"
+# For deploying account abstraction related contracts v0.6
+deployer_2_address="0x3C44CdDdB6a900fa2b585dd299e03d12FA4293BC"
+deployer_2_private_key="0x5de4111afa1a4b94908f83103eb1f1706367c2e68ca870fc3fb9a804cdab365a"
+# For deploying account abstraction related contracts v0.7
+deployer_3_address="0x90F79bf6EB2c4f870365E785982E1f101E93b906"
+deployer_3_private_key="0x7c852118294e51e653712a81e05800f419141751be58f605c371e15141b007a6"
+
```
```suggestion
# For deploying periphery contracts
deployer_1_address="0x70997970C51812dc3A010C7d01b50e0d17dc79C8"
deployer_1_private_key="0x59c6995e998f97a5a0044966f0945389dc9e86dae88c7a8412f4603b6b78690d"

# For deploying account abstraction v0.6.0 contracts
deployer_2_address="0x3C44CdDdB6a900fa2b585dd299e03d12FA4293BC"
deployer_2_private_key="0x5de4111afa1a4b94908f83103eb1f1706367c2e68ca870fc3fb9a804cdab365a"

# For deploying account abstraction v0.7.0 contracts
deployer_3_address="0x90F79bf6EB2c4f870365E785982E1f101E93b906"
deployer_3_private_key="0x7c852118294e51e653712a81e05800f419141751be58f605c371e15141b007a6"

```

pilagod commented on docker/testnet/geth_setup/entry.sh:
```diff
@@ -7,29 +7,55 @@ rpc_url="http://geth:8545"
+passkey_account_factory_address="0xBC9129Dc0487fc2E169941C75aABC539f208fb01"
+verifying_paymaster_address="0xF6168876932289D073567f347121A267095f3DD6"
+
+# Topup deployers
+cast send --rpc-url ${rpc_url} --private-key ${operator_private_key} ${deployer_1_address} --value 100ether
+cast send --rpc-url ${rpc_url} --private-key ${operator_private_key} ${deployer_2_address} --value 100ether
+cast send --rpc-url ${rpc_url} --private-key ${operator_private_key} ${deployer_3_address} --value 100ether
```
nit: Maybe we could topup them right after the declaration of deployers?

oneleo committed:
chg: Adjust the order of comments and ETH topup

oneleo committed:
chg: Adjust the contents in README.md

oneleo committed:
chg: Adjust the topup ETH order

oneleo committed:
chg: Add the git ignore list

oneleo committed:
bug: Update the EntryPoint address in rundler

oneleo committed:
Merge branch 'main' into add-aa-v0.7.0
